File: src/main/java/dev/obscuria/obscure_tooltips/client/tooltip/TooltipLabel.java
package dev.obscuria.obscure_tooltips.client.tooltip;

import com.github.bsideup.jabel.Desugar;
import com.google.gson.JsonObject;
import dev.obscuria.obscure_tooltips.client.component.TooltipComponent;
import dev.obscuria.obscure_tooltips.client.registry.TooltipRegistries;
import dev.obscuria.obscure_tooltips.client.tooltip.filter.ItemFilter;
import dev.obscuria.obscure_tooltips.client.tooltip.label.LabelProvider;
import net.minecraft.item.ItemStack;
import net.minecraft.util.JsonUtils;

import javax.annotation.Nullable;

@Desugar
public record TooltipLabel(
        int priority,
        LabelProvider provider,
        ItemFilter filter
) implements Comparable<TooltipLabel> {

    public static TooltipLabel fromJson(JsonObject json) {
        final int priority = JsonUtils.getInt(json, "priority");
        final LabelProvider provider = LabelProvider.fromJson(JsonUtils.getJsonObject(json, "provider"));
        final ItemFilter filter = ItemFilter.fromJson(JsonUtils.getJsonObject(json, "filter"));
        return new TooltipLabel(priority, provider, filter);
    }

    public boolean isFor(ItemStack stack) {
        return filter.test(stack);
    }

    public TooltipComponent create(ItemStack stack) {
        return provider.create(stack);
    }

    @Override
    public int compareTo(TooltipLabel other) {
        return Integer.compare(priority, other.priority);
    }

    public static @Nullable TooltipLabel findFor(ItemStack stack) {
        for (var label : TooltipRegistries.TOOLTIP_LABEL.listElements()) {
            if (!label.isFor(stack)) continue;
            return label;
        }
        return null;
    }
}

File: common/src/main/kotlin/dev/obscuria/tooltips/client/TooltipLabel.kt
package dev.obscuria.tooltips.client

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import dev.obscuria.tooltips.client.filter.IItemFilter
import dev.obscuria.tooltips.client.label.ILabelProvider
import dev.obscuria.tooltips.registry.TooltipsRegistries
import net.minecraft.client.gui.screens.inventory.tooltip.ClientTooltipComponent
import net.minecraft.world.item.ItemStack

class TooltipLabel(
    val priority: Int,
    val provider: ILabelProvider,
    val filter: IItemFilter
) : Comparable<TooltipLabel> {

    fun isFor(stack: ItemStack): Boolean {
        return filter.test(stack)
    }

    fun bake(stack: ItemStack): ClientTooltipComponent {
        return provider.bake(stack)
    }

    override fun compareTo(other: TooltipLabel): Int {
        return priority.compareTo(other.priority)
    }

    companion object {

        val DIRECT_CODEC: Codec<TooltipLabel> = RecordCodecBuilder.create { codec ->
            codec.group(
                Codec.INT
                    .fieldOf("priority")
                    .forGetter(TooltipLabel::priority),
                ILabelProvider.CODEC
                    .fieldOf("provider")
                    .forGetter(TooltipLabel::provider),
                IItemFilter.CODEC
                    .fieldOf("filter")
                    .forGetter(TooltipLabel::filter)
            ).apply(codec, ::TooltipLabel)
        }

        fun findFor(stack: ItemStack): TooltipLabel? {
            return TooltipsRegistries.Resource.TOOLTIP_LABEL.listElements().firstOrNull { it.isFor(stack) }
        }
    }
}